src/utils/imagePuller.ts

```diff
@@ -0,0 +1,110 @@
+import * as child_process from 'child_process';
+import { Connection } from 'vscode-languageserver';
+import { WorkspaceFolderContext } from '../services/workspaceManager';
+
+export class ImagePuller {
+  private connection: Connection;
+  private context: WorkspaceFolderContext;
+  private useProgressTracker = false;
+  private _containerEngine: string;
+  private _containerImage: string;
+  private _pullPolicy: string;
+
+  constructor(
+    connection: Connection,
+    context: WorkspaceFolderContext,
+    containerEngine: string,
+    containerImage: string,
+    pullPolicy: string
+  ) {
+    this.connection = connection;
+    this.context = context;
+    this._containerEngine = containerEngine;
+    this._containerImage = containerImage;
+    this._pullPolicy = pullPolicy;
+    this.useProgressTracker =
+      !!context.clientCapabilities.window?.workDoneProgress;
+  }
+
+  public async setupImage(): Promise<boolean> {
+    let setupComplete = false;
+    const imageTag = this._containerImage.split(':', 2)[1] || 'latest';
+    const imagePresent = this.checkForImage();
+    const pullRequired = this.determinePull(imagePresent, imageTag);
+
+    let progressTracker;
+    if (this.useProgressTracker) {
+      progressTracker = await this.connection.window.createWorkDoneProgress();
+    }
+    if (pullRequired) {
+      this.connection.console.log(
+        `Pulling image '${this._containerImage}' with pull-policy '${this._pullPolicy}' and image-tag '${imageTag}'`
+      );
+
+      try {
+        const pullCommand = `${this._containerEngine} pull ${this._containerImage}`;
+        if (progressTracker) {
+          progressTracker.begin(
+            'executionEnvironment',
+            undefined,
+            `Pulling image '${this._containerImage}'`
+          );
+        }
+        child_process.execSync(pullCommand, {
+          encoding: 'utf-8',
+        });
+        this.connection.window.showInformationMessage(
+          `Container image '${this._containerImage}' pull successful`
+        );
```

Comment:
This could also use `this.connection.console` instead

Reply:
Fixed